kbmd/kbmd.h

```diff
@@ -0,0 +1,200 @@
+int kbmd_door_setup(const char *);
+
+void kbmd_return(errf_t *restrict, struct nvlist *restrict) __NORETURN;
+uid_t req_uid(void);
+pid_t req_pid(void);
+
+void dispatch_request(struct nvlist *);
+errf_t *kbmd_zpool_create(const char *, const uint8_t *,
+    const struct ebox_tpl *, const recovery_token_t *, nvlist_t *);
+void kbmd_recover_init(int);
+void kbmd_recover_start(struct nvlist *);
+void kbmd_recover_resp(struct nvlist *);
+void kbmd_update_recovery(struct nvlist *);
+void kbmd_list_recovery(struct nvlist *);
+void kbmd_set_syspool(struct nvlist *);
+void kbmd_set_systoken(struct nvlist *);
+
+errf_t *
+template_hash(const struct ebox_tpl *, uint8_t **, size_t *);
+
+errf_t *ezfs_open(struct libzfs_handle *, const char *, int,
+    struct zfs_handle **);
+
+errf_t *get_dataset_status(const char *, boolean_t *, boolean_t *);
+errf_t *get_template(kbmd_token_t *, struct ebox_tpl **);
+errf_t *create_template(kbmd_token_t *restrict, const struct ebox_tpl *,
+    struct ebox_tpl **restrict);
+
+errf_t *set_box_name(struct ebox *, const char *);
+errf_t *kbmd_get_ebox(const char *restrict, boolean_t, struct ebox **restrict);
+errf_t *get_ebox_common(struct zfs_handle *restrict, boolean_t,
+    struct ebox **restrict);
+errf_t *kbmd_put_ebox(struct ebox *, boolean_t);
+errf_t *ebox_to_str(struct ebox *restrct, char **restrict);
+
+void kbmd_event_init(int);
+void kbmd_event_fini(void);
+errf_t *kbmd_watch_pid(pid_t, void (*)(pid_t, void *), void *);
+errf_t *kbmd_unwatch_pid(pid_t);
+
+/* piv.c */
+errf_t *kbmd_find_byguid(const uint8_t *, size_t, kbmd_token_t **);
+errf_t *kbmd_find_byslot(enum piv_slotid, const struct sshkey *,
+    kbmd_token_t **);
+errf_t *kbmd_get_slot(kbmd_token_t *restrict, enum piv_slotid slotid,
+    struct piv_slot **restrict);
+errf_t *kbmd_assert_pin(kbmd_token_t *);
+errf_t *kbmd_verify_pin(kbmd_token_t *);
+errf_t *kbmd_auth_pivtoken(kbmd_token_t *restrict, struct sshkey *restrict);
+errf_t *kbmd_setup_token(kbmd_token_t **restrict);
+void kbmd_set_token(kbmd_token_t *);
+void kbmd_token_free(kbmd_token_t *);
+errf_t *set_piv_rtoken(kbmd_token_t *, const recovery_token_t *);
+
+const char *piv_pin_str(enum piv_pin pin_type);
+
+/* plugin.c */
+errf_t *kbmd_get_pin(const uint8_t guid[restrict], struct custr **restrict);
+errf_t *register_pivtoken(kbmd_token_t *restrict,
+    struct ebox_tpl **restrict);
+errf_t *replace_pivtoken(const uint8_t [], const recovery_token_t *,
+    kbmd_token_t *restrict , struct ebox_tpl **restrict);
+errf_t *new_recovery_token(kbmd_token_t *restrict);
+errf_t *post_recovery_config_update(void);
+void load_plugin(void);
+
+/* box.c */
+errf_t *kbmd_unlock_ebox(struct ebox *restrict, struct kbmd_token **restrict);
+errf_t *kbmd_rotate_zfs_ebox(const char *);
+errf_t *kbmd_create_ebox(kbmd_token_t *restrict, const struct ebox_tpl *,
+    const char *, uint8_t **restrict, size_t *restrict, struct ebox **restrict);
+
+#define	FOREACH_STOP ((errf_t *)(uintptr_t)-1)
```

Comment:
It's a little trickier since errf_t is essentially an opaque type from pivy, but could maybe something like:
```
#define FOREACH_STOP(e) (errf_caused_by((e), "IterationStopped")
```
Then just either return an new errf_t for that, or maybe just return a static one.

Reply:
Done

Reply:
confirmed.